### PR TITLE
Translate large lookup tables to static data

### PR DESCRIPTION
**Pull request by gezalore**
How do you feel about something like this for #2925. Reusing the constants saves 20% of the output C++ in my design (and similar GCC time), and it saves like half a MB of executable code. I will add tests if otherwise OK.

### CODE REVIEW

File: Changes [hunk 1]
```diff
@@ -58,6 +58,7 @@ Verilator 4.202 2021-04-24
 * Fix V3Premit infinite loop on always read-and-write (#2898). [Raynard Qiao]
 * Fix VPI packed vectors (#2900). [Todd Strader]
 * Fix VPI public interface parameters (#2901). [Todd Strader]
+* Translate large lookup tables to static data (#2925). [Geza Lore]
```
Comment (wsnyder) [inline Changes:61]:
```suggestion
* Optimize large lookup tables to static data (#2925). [Geza Lore]
```

And move above all the "Fix" lines please.

File: src/V3Number.cpp [hunk 2]
```diff
@@ -1011,6 +1011,17 @@ uint32_t V3Number::mostSetBitP1() const {
     }
     return 0;
 }
+
+bool V3Number::eqTwoState(const V3Number& rhs) const {
+    // This is supposed to be fast
+#ifdef VL_DEBUG  // Otherwise slow
```
Comment (wsnyder) [inline src/V3Number.cpp:1017]: Did you measure this as slow?   It is rare to run with debug so we might be missing an assert that e.g. fires on the regressions.

File: src/V3EmitC.cpp [hunk 3]
```diff
@@ -1814,13 +1814,22 @@ class EmitCImp final : EmitCStmts {
...
+                        out += varp->nameProtect() + suffix + "[" + cvtToStr(w) + "] = ";
+                        out += cvtToStr(constp->num().edataWord(w)) + "U;\n";
+                    }
                 } else {
-                    out += "VL_RAND_RESET_W(";
+                    if (zeroit) {
+                        out += "VL_ZERO_RESET_W(";
+                    } else {
+                        out += "VL_RAND_RESET_W(";
+                    }
```
Comment (wsnyder) [inline src/V3EmitC.cpp:1829]:
```suggestion
                    out += zeroit ? "VL_ZERO_RESET_W(" : "VL_RAND_RESET_W(";
```

File: src/V3Premit.cpp [hunk 4]
```diff
@@ -173,16 +194,50 @@ class PremitVisitor final : public AstNVisitor {
         AstNRelinker linker;
         nodep->unlinkFrBack(&linker);
 
-        AstVar* varp = getBlockTemp(nodep);
+        AstVar* varp;
+
```
Comment (wsnyder) [inline src/V3Premit.cpp:198]:
```suggestion
        AstVar* varp = nullptr;

```

I realize right now it's always assigned, but perhaps we make a change someday.

**Review by wsnyder — COMMENTED** [review]
You're a fast worker ;)

Conceptually fine.  I think the big problem is the hashing is slow and not stable (compares pointers).  A stable key would use V3Hash to compare names or otherwise.

However instead have a per-module data structure that holds a V3Hashed structure. (This can be accessed from e.g. a function that calls m_modp->user3p() to get a pointer to the V3Hashed structure for that module, and if that is retrieves null makes a new V3Hashed and saves it in m_modp->user3p.)  When creating a new constant you'd then call V3Hashed::findDuplicate on the constant.

**Comment by wsnyder** [comment]
Also please add a V3Stat that counts how many combines are done through hashing, and have at least one test.pl file that checks this counter is non-zero.   (So if we ever break the hashing we'll fail the test.)

File: src/V3EmitC.cpp (hunk 3 above)
Comment (gezalore) [reply src/V3EmitC.cpp:1829]: This is existing code that's not related to my change, but ok

**Comment by gezalore** [comment]
> I think the big problem is the hashing is slow

The hash is 10 machine instruction, which distributes different widths to different bins, and the equals is the fastest algorithm I know of to compare two arrays for value equality, not sure why you think this is slow. It's immesurable on my design which has tons of these constants, I thoght about this.

> and not stable (compares pointers).

The only pointer involved is the parent module m_modp, which persists throughout the transformation in that pass, and we do want different bins for different modules, so pointer comparison is quite adequate and is stable. Furthermore, no ordering depends on the contents of the unordered_map (hence unordered) as the variable allocation order is in traversal order, which is deterministic. I thought about this too. All this actually is moot, because modules are not nested, so I can just clear the hash table on module entry and not needing to hash it.

File: src/V3Number.cpp (hunk 2 above)
Comment (gezalore) [reply src/V3Number.cpp:1017]: Well, it's only constant overhead I guess, so it matters less with a good hash function.

Comment (gezalore) [reply src/V3Number.cpp:1017]: Though if I load the 2 m_valueX arrays I might as well compare them and make this work 4state

Comment (gezalore) [reply src/V3Number.cpp:1017]: or drop this function and use an improved isCaseEq instead

**Comment by gezalore** [comment]
Regarding the V3Hashed, see the latest commit. All this really saves is 1 computation of the 10 instruction node hash per insertion (which only happens when a unique constant is first encountered), in exchange it's less type safe and more opaque. Wiht the improvement to isCaseEq, it's the same speed as the std::unordered_map.

**Review by wsnyder — COMMENTED** [review]
I'm good with this.  Can you add the V3Stat and test for that stat please?  What is the performance impact to Swerv in the extended tests?

**Comment by gezalore** [comment]
Done. None of the SweRV cores have big enough constants for this to trigger, so no impact there.

**Comment by wsnyder** [comment]
Good to go from my side.  Please do a pull from master first just to check that the new sanitizer doesn't find anything, then you can squash merge on pass.

**Pull request merged**